File: backend/app/schemas/device.py
from datetime import datetime

from typing import Literal

from pydantic import BaseModel, Field


class PartitionInfo(BaseModel):
    name: str
    device_path: str
    filesystem: str | None = None
    label: str | None = None
    uuid: str | None = None
    capacity_bytes: int = Field(ge=0)
    mount_points: list[str] = Field(default_factory=list)


class DeviceSummary(BaseModel):
    id: str = Field(description="Server-generated identifier; never a device path")
    device_path: str = Field(description="Read-only identity detail; clients must not submit this value")
    vendor: str | None = None
    model: str | None = None
    serial: str | None = None
    usb_version: str | None = None
    manufacturer: str | None = None
    device_class: str | None = None
    capacity_bytes: int = Field(ge=0)
    filesystem: str | None = None
    mount_point: str | None = None
    removable: bool
    read_only: bool
    transport: str | None = None
    device_type: str = Field(default="UNKNOWN", description="USB_STORAGE / INTERNAL_STORAGE / REMOVABLE_STORAGE / NVME / SATA / UNKNOWN")
    system_disk: bool = False
    detected_at: datetime


class DeviceDetail(DeviceSummary):
    partitions: list[PartitionInfo] = Field(default_factory=list)
    analysis_available: bool = True


class DeviceListResponse(BaseModel):
    devices: list[DeviceSummary]
    refreshed_at: datetime
    warning: str | None = None  # e.g. "Not running as Administrator — device detection limited"


class DeviceAnalysis(BaseModel):
    device_id: str
    filesystem: str | None = None
    total_bytes: int = Field(ge=0)
    used_bytes: int | None = Field(default=None, ge=0)
    free_bytes: int | None = Field(default=None, ge=0)
    file_count: int | None = Field(default=None, ge=0)
    directory_count: int | None = Field(default=None, ge=0)
    extension_counts: dict[str, int] | None = Field(default=None)
    status: str
    message: str


class FileEntry(BaseModel):
    """Metadata captured without opening or modifying a source file."""

    name: str
    path: str = Field(description="Path relative to the USB mount, never a host path")
    parent_path: str
    kind: Literal["file", "directory", "symlink", "other"]
    size_bytes: int | None = Field(default=None, ge=0)
    modified_at: datetime | None = None
    created_at: datetime | None = None
    accessed_at: datetime | None = None
    extension: str | None = None
    hidden: bool = False


class FileBrowserResponse(BaseModel):
    device_id: str
    current_path: str
    parent_path: str | None = None
    entries: list[FileEntry] = Field(default_factory=list)
    total: int = Field(ge=0)
    status: Literal["completed", "unavailable", "not_found"]
    message: str

File: frontend/src/types/device.ts
export interface PartitionInfo {
  name: string;
  device_path: string;
  filesystem: string | null;
  label: string | null;
  uuid: string | null;
  capacity_bytes: number;
  mount_points: string[];
}

export interface UsbDevice {
  id: string;
  device_path: string;
  vendor: string | null;
  model: string | null;
  serial: string | null;
  usb_version?: string | null;
  manufacturer?: string | null;
  device_class?: string | null;
  capacity_bytes?: number;
  size_bytes?: number;
  filesystem?: string | null;
  mount_point?: string | null;
  removable?: boolean;
  read_only?: boolean;
  transport?: string | null;
  detected_at?: string;
  device_type?: string;          // USB_STORAGE / NVME / SATA / SD_CARD / INTERNAL_STORAGE / etc.
  system_disk?: boolean;         // true if this is the OS boot disk
  is_protected?: boolean;
  is_usb?: boolean;
}

export interface UsbDeviceDetail extends UsbDevice {
  partitions: PartitionInfo[];
  system_disk: boolean;
  analysis_available: boolean;
}

export interface DeviceListResponse {
  devices: UsbDeviceDetail[];
  refreshed_at: string;
  demo_mode?: boolean; // kept optional for compat while migrating
  warning?: string | null;  // set when backend lacks admin privileges
}

export interface DeviceAnalysis {
  device_id: string;
  filesystem: string | null;
  total_bytes: number;
  used_bytes: number | null;
  free_bytes: number | null;
  file_count: number | null;
  directory_count: number | null;
  extension_counts?: Record<string, number> | null;
  status: string;
  message: string;
}

export type FileKind = "file" | "directory" | "symlink" | "other";

export interface FileEntry {
  name: string;
  path: string;
  parent_path: string;
  kind: FileKind;
  size_bytes: number | null;
  modified_at: string | null;
  created_at: string | null;
  accessed_at: string | null;
  extension: string | null;
  hidden: boolean;
}

export interface FileBrowserResponse {
  device_id: string;
  current_path: string;
  parent_path: string | null;
  entries: FileEntry[];
  total: number;
  status: "completed" | "unavailable" | "not_found";
  message: string;
}

export interface EvidenceRecord {
  id: string;
  case_id: string;
  path: string;
  sha256: string | null;
  created_at: string;
  operation_id: string;
}

export interface EvidenceListResponse {
  device_id: string;
  evidence_records: EvidenceRecord[];
}

export interface DeletedFileRecord {
  inode: string;
  filename: string;
  size_bytes: number | null;
  recoverable: boolean;
}

export interface ForensicScanResponse {
  device_id: string;
  deleted_files: DeletedFileRecord[];
  status: string;
  message: string;
}

export interface OffsetVerification {
  offset_bytes: number;
  passed: boolean;
}

export interface VerificationResult {
  device_id: string;
  status: string;
  message: string;
  is_zeroed: boolean;
  sampled_offsets: OffsetVerification[];
}
